### PR TITLE
Fix: addShape handling percentage attributes

### PR DESCRIPTION
**Pull request by tkez**
Before passing into _Path.rect, convert the percentage attributes into value, with _currentParentElement assigned in _readSubtree

### CODE REVIEW

**Comment by tkez** [comment]
Not sure if it provide any help, but it did solve the issue I'm encounter with.

**Comment by dnfield** [comment]
Thanks!

Can we add some tests for this?  Perhaps a new golden SVG for it?

**Comment by tkez** [comment]
I really love to do that, but even with original version of flutter_svg, I'm having below failed test when I run the golden_svg_test.
side note: I'm on windows machine

> If that element is meant to be visible, the `display` or `visibility` attributes should be removed.
> If that element is not meant to be visible, it would be better to remove it from the SVG file.
> SVG Warning: Discarding:
> 
>   <rect visibility="hidden" width="100" height="50" x="10" y="70" fill="rgb(255, 0, 0)" stroke="blue" stroke-opacity=".3" rx="20" ry="15"/>
> 
> and any children it has since it is not visible.
> If that element is meant to be visible, the `display` or `visibility` attributes should be removed.
> If that element is not meant to be visible, it would be better to remove it from the SVG file.
> package:test_api                expect
> test\golden_svg_test.dart 40:7  main.<fn>
> ===== asynchronous gap ===========================
> dart:async                      _AsyncAwaitCompleter.completeError
> test\golden_svg_test.dart       main.<fn>
> ===== asynchronous gap ===========================
> dart:async                      _asyncThenWrapperHelper
> test\golden_svg_test.dart       main.<fn>
> 
> Expected: equals [
>             137,
>             80,
>             78,
>             71,
>             13,
>             10,
>             26,
>             10,
>             0,
>             0,
>             0,
>             13,
>             73,
>             72,
>             68,
>             82,
>             0,
>             0,
>             0,
>             200,
>             0,
>             0,
>             0,
>             200,
>             ...
>           ] ordered
>   Actual: [
>             137,
>             80,
>             78,
>             71,
>             13,
>             10,
>             26,
>             10,
>             0,
>             0,
>             0,
>             13,
>             73,
>             72,
>             68,
>             82,
>             0,
>             0,
>             0,
>             200,
>             0,
>             0,
>             0,
>             200,
>             ...
>           ]
>    Which: was <21> instead of <22> at location [64]

**Comment by dnfield** [comment]
Ahh you probably can't properly generate the images on Windows, or a Skia roll has changed the way this happens in the framework...  If you add the SVG I can generate the goldne for you.

**Comment by tkez** [comment]
oh..ok.
Below is the svg that I initially having issue with the %
Is this the one you interested in?

`<svg xmlns="http://www.w3.org/2000/svg" width="150" height="50"><rect width="100%" height="100%" fill="#f5f7fa"/><path fill="#e584b5" d="M27.38 12.75L26.93 13.45L25.60 13.49L25.49 13.38Q25.10 13.66 24.69 13.49L24.56 13.36L19.47 13.94L18.95 14.78L18.87 14.70Q19.37 17.42 26.25 17.31L26.20 17.27L26.27 17.33Q24.41 20.13 22.19 21.21L22.14 21.16L22.11 23.63L23.40 24.45L23.55 24.59Q26.42 23.09 29.37 23.37L29.48 23.49L29.48 24.38L29.46 24.36Q30.68 25.03 29.43 27.98L29.48 28.03L28.22 28.14L28.21 28.13Q25.99 31.71 22.07 31.99L22.17 32.10L22.19 32.11Q20.85 33.04 19.19 32.97L19.10 32.88L19.22 33.01Q18.37 33.91 18.44 35.16L18.57 35.29L19.62 36.36L19.66 36.40Q20.78 36.14 21.62 35.62L21.71 35.71L24.02 35.74L24.32 34.68L24.49 34.85Q28.44 33.65 30.14 30.52L30.06 30.44L30.14 30.52Q34.29 27.17 30.82 21.20L30.66 21.04L30.81 21.19Q29.12 21.03 28.09 20.29L28.11 20.31L27.94 20.14Q34.52 14.01 28.44 13.60L28.48 13.64L27.40 12.77Z"/><path fill="#dd54dd" d="M116.07 12.76L115.94 12.63Q114.02 13.39 112.47 13.37L112.43 13.33L112.33 13.23Q105.57 20.67 107.91 29.56L107.93 29.59L109.25 29.78L109.33 29.86Q109.65 30.93 108.54 30.75L108.47 30.68L108.49 31.83L110.44 32.39L110.37 32.32Q109.52 33.68 113.16 36.22L113.19 36.25L115.47 36.29L117.41 36.27L118.28 36.29L118.32 36.33Q118.76 33.60 121.55 33.94L121.68 34.07L121.84 32.55L122.28 32.45L123.00 31.77L122.82 30.17L122.81 30.17Q124.54 27.35 124.16 24.51L124.19 24.53L125.09 24.31L125.00 23.37L124.94 23.31Q124.74 23.00 124.77 22.58L124.62 22.44L124.81 21.78L124.75 21.71Q123.76 19.93 123.78 18.51L123.92 18.66L123.94 18.67Q122.18 16.36 121.93 14.41L121.88 14.36L121.87 14.36Q119.99 14.24 118.93 13.10L118.78 12.95L118.87 13.04Q117.32 13.08 116.02 12.71L116.12 12.80ZM113.60 17.30L113.58 17.28Q119.38 16.70 120.42 21.87L120.47 21.92L120.45 21.90Q121.03 22.32 121.45 22.06L121.36 21.97L121.26 21.87Q122.30 27.56 119.34 31.75L119.35 31.76L119.37 31.77Q117.16 33.73 114.87 32.32L115.00 32.45L114.95 32.40Q114.16 33.52 113.49 32.34L113.43 32.29L113.49 32.35Q111.30 32.30 111.19 30.05L111.17 30.03L111.26 30.12Q109.95 29.96 109.07 29.32L109.10 29.34L109.65 28.49L109.52 28.36Q108.44 24.20 110.38 21.94L110.40 21.97L110.51 22.07Q110.39 18.34 113.41 17.11L113.50 17.20Z"/><path fill="#eb734b" d="M54.31 12.63L54.33 12.65Q47.09 15.30 48.04 22.60L47.97 22.53L48.06 22.61Q49.61 23.93 50.14 25.64L50.17 25.66L50.09 25.58Q55.09 26.78 58.80 24.89L58.80 24.89L59.67 26.21L59.49 26.04Q57.04 32.47 51.15 33.47L51.19 33.52L51.00 33.32Q50.00 34.79 50.23 36.32L50.14 36.23L50.22 36.31Q56.77 35.39 58.77 29.56L58.94 29.73L58.90 29.69Q61.41 27.80 60.77 24.97L60.73 24.92L61.17 25.14L61.29 25.25Q62.53 21.37 60.77 18.61L60.75 18.58L60.80 18.64Q60.54 13.04 54.31 12.63L54.47 12.79ZM53.04 16.31L53.01 16.27Q54.41 16.37 55.59 16.26L55.67 16.33L55.63 16.30Q56.39 17.48 57.92 17.41L57.86 17.35L57.89 17.39Q61.56 25.52 52.26 23.76L52.31 23.80L52.20 23.69Q51.31 22.04 50.31 21.10L50.44 21.23L50.32 21.11Q50.08 17.34 52.97 16.23L52.94 16.20Z"/><path d="M3 28 C70 15,72 7,142 34" stroke="#6d89dd" fill="none"/><path fill="#a570db" d="M86.46 12.64L86.00 13.31L84.78 13.47L84.69 13.38Q84.19 13.55 83.78 13.38L83.76 13.36L78.50 13.77L78.18 14.81L78.14 14.77Q78.49 17.34 85.36 17.23L85.39 17.26L85.34 17.21Q83.61 20.13 81.39 21.21L81.34 21.16L81.24 23.56L82.66 24.51L82.65 24.50Q85.72 23.19 88.67 23.48L88.57 23.37L88.69 24.39L88.68 24.39Q89.92 25.07 88.67 28.02L88.65 28.00L87.35 28.08L87.47 28.20Q85.31 31.83 81.39 32.11L81.36 32.08L81.38 32.11Q79.97 32.96 78.31 32.89L78.42 33.00L78.36 32.94Q77.72 34.05 77.78 35.30L77.75 35.27L78.73 36.26L78.88 36.42Q79.93 36.09 80.77 35.57L80.82 35.62L83.19 35.71L83.66 34.82L83.53 34.69Q87.51 33.52 89.22 30.39L89.28 30.46L89.35 30.53Q93.49 27.16 90.02 21.20L89.94 21.11L90.04 21.22Q88.24 20.95 87.21 20.21L87.12 20.13L87.25 20.26Q93.72 14.01 87.65 13.60L87.67 13.63L86.63 12.81Z"/></svg>`

File: lib/src/svg/parser_state.dart [hunk 1]
```diff
@@ -53,19 +53,18 @@ Offset _parseCurrentOffset(SvgParserState parserState, Offset lastOffset) {
     x != null
         ? parseDouble(x)
         : parseDouble(parserState.attribute('dx', def: '0')) +
-            (lastOffset?.dx ?? 0),
+        (lastOffset?.dx ?? 0),
```
Comment (dnfield) [inline lib/src/svg/parser_state.dart]: Something seems to have gone off with the formatting here...

**Comment by dnfield** [comment]
Looking at this a bit more closely now.  

This should be doable, but I think the current approach has a couple flaws - to figure out a percentage length, we need to know the bounding box.  That bounding box may be established by any ancestor, not necessarily the immediate parent, and is specified here: https://www.w3.org/TR/SVG11/coords.html#ObjectBoundingBox

We should probably be supporting that bounding box model (which, in your example case, is just the viewbox) properly and passing it along through the parser state somehow.  Once we have that, we could determine the percentage of the bounding box dimension that we want.

Does that make sense to you?

**Comment by tkez** [comment]
You most certainly right :)

**Comment by tkez** [comment]
I think I would look more into it if I find time to do so.
It would be much appreciate if can get help from getting those svg test case :)

**Comment by dnfield** [comment]
I'll look at this at some point if you don't get to it.  Just not sure when I'll get to it yet.

**Comment by tkez** [comment]
Just notice how the xml parse through the elements.
Now only will process width and height attribute, and will get parent attributes from as top as necessary.
Also added below svg to the golden svg assets.

`<svg xmlns="http://www.w3.org/2000/svg" width="100%" viewBox="0 0 3945.8 2400">
    
    <svg id="birdwrapper" x="698" y="1219" width="1055" height="641" style="overflow: visible;">
        
        <rect x="0" y="0" width="100%" height="100%" fill="none" stroke="#000" stroke-width="10"></rect>
        
        <svg id="bird" style="overflow: visible;" viewBox="0 0 1055 641" x="50%" y="50%">
            
            <rect x="0" y="0" width="100%" height="100%" fill="none" stroke="#f00" stroke-width="10"></rect>
            
            <g>
                <line x1="528.9" y1="671.3" x2="497.8" y2="556.3" fill="none" stroke="#fdb913" stroke-miterlimit="10" stroke-width="10"/>
                <polyline points="545.5 655.9 523.8 651.9 508.3 666.3" fill="none" stroke="#fdb913" stroke-miterlimit="10" stroke-width="10"/>
                <line x1="464.5" y1="680.2" x2="437.8" y2="581.3" fill="none" stroke="#fdb913" stroke-miterlimit="10" stroke-width="10"/>
                <polyline points="481.1 664.8 459.3 660.8 443.9 675.1" fill="none" stroke="#fdb913" stroke-miterlimit="10" stroke-width="10"/>
            </g>
            <path d="M544,234.9c79.1,0,110.5,65.2,86.8,160.6-24.2,97.3-109.1,204.3-293.2,210.2-59,1.9-117.3-28.4-117.3-28.4S339.9,234.9,544,234.9Z" fill="#ed1c24"/>
            <polyline points="637.7 321 705.8 347.5 637.7 355.6" fill="#fdb913"/>
            <path d="M220.3,577.4a269.2,269.2,0,0,0,30.8,12.9c59.5-142.2,168.5-317.8,378.4-303.1-13.6-33-42.5-52.3-85.6-52.3C339.9,234.9,220.3,577.4,220.3,577.4Z" fill="#d71920"/>
            <path d="M357.1,604.7c134.7-10,212.8-76.1,251.2-149.7C496.2,459.5,407.8,539.9,357.1,604.7Z" fill="#f15b40"/>
            <path d="M251.6,321.2c-53-3-92.9,20.5-95,52.1S191,432.7,244,435.7c88.9,5,167-48,167-48S340.4,326.3,251.6,321.2Z" fill="#f15b40"/>
            <path d="M236.2,384.8c-50,16.6-76.7,52.7-65.2,82.5s56.9,42.1,106.9,25.5C361.7,465,411,387.7,411,387.7S320,357,236.2,384.8Z" fill="#f15b40"/>
            <circle cx="574.6" cy="299.8" r="21.2" fill="#231f20"/>
        </svg>
    </svg>
</svg>`

**Comment by dnfield** [comment]
I'm going to close this - it's out of date. If you want to attempt to complete this patch, please feel free to ping me here or open a new PR.

**Pull request closed without merging**